packages/product/src/services/product-module-service.ts

```diff
@@ -197,58 +198,158 @@ export default class ProductModuleService<
     return joinerConfig
   }
 
-  async createVariants(
+  // TODO: Add options validation, among other things
+  createVariants(
     data: ProductTypes.CreateProductVariantDTO[],
+    sharedContext?: Context
+  ): Promise<ProductTypes.ProductVariantDTO[]>
+  createVariants(
+    data: ProductTypes.CreateProductVariantDTO,
+    sharedContext?: Context
+  ): Promise<ProductTypes.ProductVariantDTO>
+
+  @InjectManager("baseRepository_")
+  async createVariants(
+    data:
+      | ProductTypes.CreateProductVariantDTO[]
+      | ProductTypes.CreateProductVariantDTO,
     @MedusaContext() sharedContext: Context = {}
-  ): Promise<ProductTypes.ProductVariantDTO[]> {
-    const productVariantsWithOptions = await promiseAll(
-      data.map(async (variant) => {
-        if (!isPresent(variant.options)) {
-          return variant
-        }
+  ): Promise<
+    ProductTypes.ProductVariantDTO[] | ProductTypes.ProductVariantDTO
+  > {
+    const input = Array.isArray(data) ? data : [data]
 
-        const productOptions = await this.productOptionService_.list(
-          { product_id: variant.product_id },
-          {
-            take: null,
-          },
-          sharedContext
-        )
+    const variants = await this.createVariants_(input, sharedContext)
 
-        return ProductModuleService.assignOptionsToVariants(
-          [variant],
-          productOptions
-        )[0]
-      })
+    const createdVariants = await this.baseRepository_.serialize<
+      ProductTypes.ProductVariantDTO[]
+    >(variants, { populate: true })
+
+    return Array.isArray(data) ? createdVariants : createdVariants[0]
+  }
+
+  async createVariants_(
+    data: ProductTypes.CreateProductVariantDTO[],
+    @MedusaContext() sharedContext: Context = {}
+  ): Promise<ProductVariant[]> {
+    if (data.some((v) => !v.product_id)) {
+      throw new MedusaError(
+        MedusaError.Types.INVALID_DATA,
+        "Tried to create variants without specifying a product_id"
+      )
+    }
+
+    const productOptions = await this.productOptionService_.list(
+      {
+        product_id: uniqBy(data, "product_id").map(
```

Comment:
```suggestion
        product_id: [...new Set<string>(data.map(d => d.product_id))]
```

Reply:
not sure this util uniqBy is required, also, a single loop is necessary, which also simplify the readability and cyclomatic complexity, even if we dont really care here